src/briefcase/platforms/macOS/__init__.py

```diff
@@ -160,37 +181,64 @@ def select_identity(self, identity=None):
             print()
             selection = select_option(identities, input=self.input)
             identity = identities[selection]
-            print("selected", identity)
 
         return identity
 
-    def sign(self, path, entitlements, identity):
+    def sign(self, path, identity, entitlements=None):
         """
         Code sign a file.
 
         :param path: The path to the file to sign.
-        :param entitlements: The path to the entitlements file to use.
         :param identity: The code signing identity to use. Either the 40-digit
             hex checksum, or the string name of the identity.
+        :param entitlements: The path to the entitlements file to use.
         """
+        options = 'runtime' if identity != '-' else None
+        process_command = [
+            'codesign',
+            os.fsdecode(path),
+            '--sign', identity,
+            '--force',
+        ]
+        if entitlements:
+            process_command.append('--entitlements')
+            process_command.append(os.fsdecode(entitlements))
+        if options:
+            process_command.append('--options')
+            process_command.append(options)
+
         try:
-            print("Signing", path)
+            print(f"Signing {Path(path).relative_to(self.base_path)}")
             self.subprocess.run(
-                [
-                    'codesign',
-                    '--sign', identity,
-                    '--entitlements', os.fsdecode(entitlements),
-                    '--deep', os.fsdecode(path),
-                    '--force',
-                    '--options', 'runtime',
-                ],
+                process_command,
+                stderr=subprocess.PIPE,
                 check=True,
             )
-        except subprocess.CalledProcessError:
-            print()
-            raise BriefcaseCommandError(
-                "Unable to code sign {path}.".format(path=path)
-            )
+        except subprocess.CalledProcessError as e:
+            errors = e.stderr.decode('utf-8', errors='replace')
+            if 'not signed at all' in errors:
```

Comment:
@dreavjr Are you able to provide an example of how to manufacture this message? I'm not sure if I can do anything to manufacture a test for this, but at the very least, I'd like to reproduce it manually.

Reply:
I'm away from my main machine right now, so I'm cannot give you an exact procedure. This error happened to me while processing Qt for Python 6 due to some violations the package has on the strictly hierarchical structure Apple expects. Thus, signing from the inside out was not enough to warrant that one of the binaries inside Qt for Python was correctly signed, and that one binary had to be signed with the "--deep" option. 

As I understood, this kind of situation is supposed to be rare, so we should not use "--deep" preemptively on everything, because that would perform a huge amount of redundant work.

Reply:
Ok - I've been able to generate this error packaging a PySide6 app.